godot-core/src/builtin/script.rs

```diff
@@ -0,0 +1,480 @@
+
+        let property_list: Vec<_> = (*instance)
+            .inner
+            .borrow()
+            .property_list()
+            .iter()
+            .map(|prop| prop.property_sys())
+            .collect();
+
+        *r_count = property_list.len() as u32;
+
+        let ptr = Box::into_raw(property_list.into_boxed_slice());
+
+        (*ptr).as_ptr()
+    }
+
+    pub(super) unsafe extern "C" fn get_method_list_func<T: ScriptInstance>(
+        p_instance: GDExtensionScriptInstanceDataPtr,
+        r_count: *mut u32,
+    ) -> *const GDExtensionMethodInfo {
+        let instance = p_instance as *const ScriptInstanceData<T>;
+
+        let method_list: Vec<_> = (*instance)
+            .inner
+            .borrow()
+            .method_list()
+            .iter()
+            .map(|method| method.method_sys())
+            .collect();
+
+        *r_count = method_list.len() as u32;
+
+        let ptr = Box::into_raw(method_list.into_boxed_slice());
+
+        (*ptr).as_ptr()
+    }
+
+    pub(super) unsafe extern "C" fn free_property_list_func<T: ScriptInstance>(
+        p_instance: GDExtensionScriptInstanceDataPtr,
+        p_prop_info: *const GDExtensionPropertyInfo,
+    ) {
+        let instance = p_instance as *const ScriptInstanceData<T>;
+        let length = (*instance).inner.borrow().property_list().len();
+
+        let vec = Vec::from_raw_parts(p_prop_info as *mut GDExtensionPropertyInfo, length, length);
+
+        drop(vec);
+    }
+
+    pub(super) unsafe extern "C" fn call_func<T: ScriptInstance>(
+        p_self: GDExtensionScriptInstanceDataPtr,
+        p_method: GDExtensionConstStringNamePtr,
+        p_args: *const GDExtensionConstVariantPtr,
+        p_argument_count: GDExtensionInt,
+        r_return: GDExtensionVariantPtr,
+        r_error: *mut GDExtensionCallError,
+    ) {
+        let instance = p_self as *mut ScriptInstanceData<T>;
+        let method =
+            StringName::from_arg_ptr(p_method as GDExtensionTypePtr, PtrcallType::Standard);
+
+        let args: Vec<_> = slice::from_raw_parts(p_args, p_argument_count as usize)
+            .iter()
+            .map(|ptr| &*Variant::ptr_from_sys(force_mut_ptr(*ptr)))
+            .collect();
+
+        match (*instance)
+            .inner
+            .borrow_mut()
+            .call(method.clone(), args.as_slice())
+        {
+            Ok(ret) => {
+                ret.move_return_ptr(r_return as GDExtensionTypePtr, PtrcallType::Virtual);
+                (*r_error).error = GDEXTENSION_CALL_OK;
+            }
+
+            Err(err) => (*r_error).error = err,
+        };
+    }
+
+    pub(super) unsafe extern "C" fn get_script_func<T: ScriptInstance>(
+        p_instance: GDExtensionScriptInstanceDataPtr,
+    ) -> GDExtensionObjectPtr {
+        let instance = p_instance as *const ScriptInstanceData<T>;
+
+        let script = (*instance).inner.borrow().get_script();
+        let ptr = script.obj_sys();
+
+        mem::forget(script.into_ffi());
```

Comment:
here it'd also be nice to know why we leak this memory.

Reply:
`Script` is a ref counted Godot object, and the engine wants a pointer to a script. Godot then does a reinterprete_cast into a `Ref<Script>` so I'm understanding that as that the engine is now responsible for decrementing the ref count.

Reply:
I believe that's correct yeah, it'd be good to have a comment there stating this explicitly though. And the function itself should also document that the returned pointer is an owned Ref-pointer.